core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/BinPackingNodeAllocatorService.java

```diff
@@ -645,7 +645,13 @@ public ReserveResult tryReserve(PendingAcquire acquire)
             catalogNodes.ifPresent(candidates::retainAll); // Drop non-catalog nodes, if any.
             Set<HostAddress> addresses = requirements.getAddresses();
             if (!addresses.isEmpty()) {
-                candidates = candidates.stream().filter(node -> addresses.contains(node.getHostAndPort())).collect(toImmutableList());
+                List<InternalNode> preferred = candidates.stream().filter(node -> addresses.contains(node.getHostAndPort())).collect(toImmutableList());
+                if (preferred.isEmpty() && requirements.isRemotelyAccessible()) {
```

Comment:
This seems almost equivalent to the change in `UniformNodeSelector`, but without the `optimizedLocalScheduling` check.  That seems odd.  Why is this failover unconditional here but conditional on the `optimizedLocalScheduling` guard in `UniformNodeSelector`?

Reply:
accounting for `optimizedLocalScheduling` here